pkg/maps/ctmap/ctmap.go

```diff
@@ -179,8 +181,13 @@ func InitMapInfo(tcpMaxEntries, anyMaxEntries int, v4, v6 bool) {
 		anyMaxEntries, natMaps[mapTypeIPv6AnyGlobal])
 }
 
-func init() {
-	InitMapInfo(option.CTMapEntriesGlobalTCPDefault, option.CTMapEntriesGlobalAnyDefault, true, true)
+func getMapInfo(lru bool) map[mapType]mapAttributes {
+	once.Do(func() {
+		initMapInfo(option.CTMapEntriesGlobalTCPDefault,
+			option.CTMapEntriesGlobalAnyDefault, option.Config.EnableIPv4,
+			option.Config.EnableIPv6, lru)
+	})
+	return mapInfo
```

Comment:
For reviewers: this is actually a bit fragile. We need to ensure the map is created with the correct map type to avoid breakage (i.e., regeneration sees different map type and recreates the map + empty it). Some code paths may call `getMapInfo()` with `lru=false` because the map type doesn't matter for them (e.g., when listing entries). So if one of these code paths calls `getMapInfo(false)` before the actual map creation, map info. will be initialized without LRU map types when datapath might actually expect LRU maps. This currently does not fail because map creation happens first.

I'm not quite sure how to fix that one. I can either always pass the correct value for `lru`, which may involves calling bpftool in all call sites, or I can have two sets of map info., one for map creation (map type matters) and one for other map uses (map type doesn't matter)?

Reply:
Thanks for pointing this out, it is indeed an odd one. Can't we detect that if is LRU as soon as possible and set that as a global variable in the `pkg/maps/ctmap`? I don't like this idea as well :-(

 Also, why are we using `option.CTMapEntriesGlobalTCPDefault` and `option.CTMapEntriesGlobalAnyDefault` to initialize the maps shouldn't we pick `option.Config.*` for each one?

Reply:
> Thanks for pointing this out, it is indeed an odd one. Can't we detect that if is LRU as soon as possible and set that as a global variable in the `pkg/maps/ctmap`? I don't like this idea as well :-(

I guess that already kinda of what I do since the map creation with LRU set/unset happens during daemon initialization (`initMaps()` is called in `NewDaemon()`). Unless you had an even earlier place in mind?

> Also, why are we using `option.CTMapEntriesGlobalTCPDefault` and `option.CTMapEntriesGlobalAnyDefault` to initialize the maps shouldn't we pick `option.Config.*` for each one?

I think this is just to pass the global values to `initMapInfo()` which already selects the appropriate value depending on global vs. local (it sets `mapNumEntriesLocal` for local ct. maps). To be clear, I didn't change this aspect of the code.

Reply:
I guess same thought as above, but I don't understand why now everywhere needs to be concerned about the LRU detail deep in the datapath. 

> I'm not quite sure how to fix that one. I can either always pass the correct value for lru, which may involves calling bpftool in all call sites,

FWIW we already cache this in memory after first bpftool invocation, so it shouldn't be costly. This is what `GetMapType()` is already doing.

> I guess that already kinda of what I do since the map creation with LRU set/unset happens during daemon initialization (`initMaps()` is called in `NewDaemon()`). 

What would help here? You want to know if it's possible for other paths to call into the ctmap before `initMaps()` is called? CTMap should primarily be used from userspace by the garbage collector, which is tied into a controller I believe, so we should be able to ensure that the controller isn't launched until later.  If the dependency there isn't explicit enough, one thought is we could e.g. introduce a channel that's closed when the maps are initialized, then the subsequent map access must block on the channel being closed before they can access the map.

Reply:
> FWIW we already cache this in memory after first bpftool invocation, so it shouldn't be costly. This is what `GetMapType()` is already doing.

I mostly wanted to avoid the additional code (and for tests, the need to have privileges).
Anyway, our answer above solves this concern :-)